File: python/algo/202507/algo1970.py
from typing import List

# 常用数据结构 - 7.2 并查集 - 进阶
class UnionFind:
    def __init__(self, n:int):
        self.fa = list(range(n))
        self.cc = n                 # 连通块数目
    
    def find(self, x: int) -> int:
        if self.fa[x] != x:
            self.fa[x] = self.find(self.fa[x])
        return self.fa[x]

    def merge(self, u: int, v:int) -> None:
        x, y = self.find(u), self.find(v)
        if x != y:
            self.fa[x] = y
            self.cc -= 1

class Solution:
    def latestDayToCross(self, row: int, col: int, cells: List[List[int]]) -> int:
        # 参考题解，时光倒流 - 使用并查集
        uf = UnionFind(row*col + 2)
        top, bottom = row * col, row * col + 1
        grid = [[0] * col for _ in range(row)]
        for i in range(len(cells)-1, -1, -1):   # 倒序遍历
            r, c = cells[i][0]-1, cells[i][1]-1
            grid[r][c] = 1
            for nr, nc in [(r-1,c),(r+1,c),(r,c-1),(r,c+1)]:
                if 0 <= nr < row and 0 <= nc < col and grid[nr][nc]:
                    uf.merge(r*col+c, nr*col+nc)
                if r == 0:  uf.merge(r*col + c, top)
                if r == row-1:  uf.merge(r*col + c, bottom)
            if uf.find(top) == uf.find(bottom): return i

File: cplus/leetcode/202507/algo1970.cpp
#include <bits/stdc++.h>
using namespace std;

// 常用数据结构 - 7.2  并查集 - 进阶
class UnionFind {
    vector<int> fa;
    int row, col;
    unordered_map<int, set<int>> mp;
public:
    UnionFind(int n, int r, int c): fa(n), row(r), col(c) {
        ranges::iota(fa, 0);
    }

    int find(int x) {
        if (fa[x] != x) {
            fa[x] = find(fa[x]);
        }
        return fa[x];
    }

    void merge(int from, int to) {
        int x = find(from), y = find(to);
        if (x != y) {
            fa[x] = y;
            if (mp.count(x)) mp[y].insert(mp[x].begin(), mp[x].end());
            else mp[y].insert(x%col);
            mp.erase(x);                    // 删除空集合
            mp[y].insert(y%col);
        }
    }
    
    bool is_fill(int x) {
        return mp[find(x)].size() == col;
    }
};


class Solution {
public:
    int latestDayToCross(int row, int col, vector<vector<int>>& cells) {
        // 自行解答，每天水域多一格，将水域的八个方向连接起来(阻断陆地)
        // 当水域的的元节点的列数量等于列数时，说明上下的陆地被完全阻断
        const int dirs[8][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}, {-1, -1}, {-1, 1}, {1, -1}, {1, 1}};
        vector<vector<int>> grid(row, vector<int>(col));
        UnionFind uf(row * col, row, col);   // 多出两个点，一个在最上面，一个在最下面
        
        int ans = 0;
        for (auto& cell : cells) {
            int i = cell[0]-1, j = cell[1]-1;
            grid[i][j] = 1;
            for (auto& dir : dirs) {
                int x = i + dir[0], y = j + dir[1];
                if (x >= 0 && x < row && y >= 0 && y < col && grid[x][y]) {
                    uf.merge(i * col + j, x * col + y);
                }
            }
            if (!uf.is_fill(i * col + j)) ans ++;
            else break;
        }
        return ans;
    }

};